[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1bv9Di1PmgvNHiL8Ue2LWTjekvfYXEQ8k?usp=sharing)

In [1]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [2]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [3]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

8
9
17

### Test input output files

In [4]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [5]:
%%writefile output1.txt
12

Overwriting output1.txt


In [6]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [7]:
%%writefile output2.txt
9

Overwriting output2.txt


In [8]:
!./APlusB < input1.txt

12

In [9]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [10]:
!grep -xvFf output1.txt output1.txt

In [11]:
!grep -xvFf output1.txt output2.txt

9


In [12]:
!grep -xvFf output2.txt output1.txt

12


In [13]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [14]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [15]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [16]:
!python sum.py

6 3
9


In [17]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [18]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [19]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


##### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [20]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

3
1
2
3
6


### Test 1: Small numbers

In [21]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Overwriting input0.txt


In [22]:
%%writefile output0.txt
0

Overwriting output0.txt


In [23]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [24]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [25]:
!./max_pairwise_product_full < input1.txt

90


In [26]:
%%writefile output1.txt
90

Overwriting output1.txt


In [27]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [28]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [29]:
%%writefile output2.txt
90

Overwriting output2.txt


In [30]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [31]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Overwriting input3.txt


In [32]:
200000*200000

40000000000

In [33]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [34]:
!./max_pairwise_product_full < input3.txt

1345294336


In [35]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [36]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [37]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [38]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [39]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [40]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [41]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {0..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


comparacion de salida esperada con la salida obtenida, usando tipos de dato long long para las variables y retorno de la funcion max_product


In [42]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [43]:
%%writefile input4.txt
10
200000 23 200000 0 45 1 4 6 123 32131

Overwriting input4.txt


In [44]:
200000 * 200000

40000000000

In [45]:
%%writefile output4.txt
40000000000

Overwriting output4.txt


In [46]:
%%writefile input5.txt
5
3 1 200000 45 20000

Overwriting input5.txt


In [47]:
200000 * 20000

4000000000

In [48]:
%%writefile output5.txt
4000000000

Overwriting output5.txt


In [49]:
%%writefile input6.txt
15
199999 234 0 234 3000 124511 400000 4523 7845 121314 89677 453123 523213 7453 0

Overwriting input6.txt


In [50]:
453123 * 523213

237079844199

In [51]:
%%writefile output6.txt
237079844199

Overwriting output6.txt


In [52]:
%%writefile input7.txt
2
900000 5

Overwriting input7.txt


In [53]:
900000 * 5

4500000

In [54]:
%%writefile output7.txt
4500000

Overwriting output7.txt


In [55]:
%%writefile input8.txt
19
28543 11213 7663 35730 28443 14637 19039 25441 20156 16505 30649 23923 25435 18913 29108 25409 17174 143 11733

Overwriting input8.txt


In [56]:
35730 * 30649

1095088770

In [57]:
%%writefile output8.txt
1095088770

Overwriting output8.txt


In [58]:
%%writefile input9.txt
50
59918 10147 58122 15353 69460 86114 18171 45377 79182 68269
33235 50458 37455 14672 78008 28725 41017 56580 49624 29429
23106 12679 29952 26702 84626 25185 45291 89627 38901 23801
53468 80883 68361 72442 87880 34449 17123 56210 52144 46789
62951 12690 46326 85744 11059 36021 49412 11731 21735 33212

Overwriting input9.txt


In [59]:
89627 * 87880

7876420760

In [60]:
%%writefile output9.txt
7876420760

Overwriting output9.txt


In [61]:
%%writefile input10.txt
60
16695 50605 42530 57855 24236 24347 36570 58107 25949 26939
10769 54063 66323 23606 22007 37126 51412 14381 43186 12283
34525 55947 52947 66780 14485 41951 63416 32926 27342 64352
37077 27768 27124 54245 63911 28044 12350 64340 57079 51652
38353 17289 45777 45391 22221 60764 26161 36591 12275 49150
123888 129855 116245 96360 107581 124692 133711 82847 116602 103602

Overwriting input10.txt


In [62]:
133711 * 129855

17363041905

In [63]:
%%writefile output10.txt
17363041905

Overwriting output10.txt


creacion de script para la comparacion de los casos de prueba, salida esperada con salida obtenida

In [64]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {0..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


como la respuesta obtenida despues de la ejecucion del script es vacia significa que hubo coincidencias exactas (usando las entradas de input.txt) entre las salidas esperdas y las salidas obtenidas

In [65]:
#Running tests cases. Check "/files" for "outputN.txt" where N is the Nth output for the Nth "inputN.txt" (e.g. input4.txt/output4.txt)
!chmod 744 stress_test.sh
!./stress_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [66]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [67]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [68]:
199999 * 200000

39999800000

salida esperada del max producto

In [69]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [70]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [71]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [72]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

calculo del max_producto de una lista de 200k elementos del 1 al 200k

In [73]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	7m5.296s
user	6m22.823s
sys	0m0.219s


calculo del max_producto de una lista de 200k elementos del 200k al 1

In [74]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m0.689s
user	4m58.247s
sys	0m0.084s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [75]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound

Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [76]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 ////INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);
    execints++;

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right) \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$

In [77]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

571


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$

In [78]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [79]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [80]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [81]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

601


For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [82]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [83]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt
n= 10

572


For the input11.txt
```
200000
1 2 ..  199999 200000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$



In [84]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [85]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177092


In [86]:
# Estimated time by instruction
295/200001000021


1.4749926248820013e-09

In [87]:
# Estimated number of instructions by second
1 / (295/200001000021)

677969491.5966102

# Estimate the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


inpunt11.txt lista de 11000 elementos, del 1 al 11k

In [88]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


compilamos y ejecutamos el archivo para crear el archivo input13.txt con las 11k entradas

In [89]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

ejecutamos el algoritmo para encontrar el max_producto pasando como entrada la lista de 11k elementos del archivo input13.txt

salida esperada: 11000 * 10999 = 120989000


In [90]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.846s
user	0m0.844s
sys	0m0.000s


## Python

In [91]:
%%writefile max_pairwise_product_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(n):
            if i != j:
                aux_product = l[i] * l[j]
                if aux_product > max_product:
                    max_product = aux_product
    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_py.py


cod. fuente para crear el archivo input14.txt con los valores del 1 al 11k

In [92]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("11000\n")
        for i in range(1, 11001):
            f.write(f"{i} ")

Writing WriteTestPython.py


ejecucion del cod fuente anterior para crear el archivo input14.txt

In [93]:
!python WriteTestPython.py

ejecucion del algoritmo en python con las entradas del archivo input14.txt

In [94]:
# Verify #Not surprising Python version is slower...
%%time
!python max_pairwise_product_py.py < input14.txt

120989000
CPU times: user 153 ms, sys: 25.6 ms, total: 179 ms
Wall time: 21.2 s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [95]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [96]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound

Writing max_pairwise_product_long.cpp


In [97]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m22.142s
user	2m20.817s
sys	0m0.053s


In [98]:
%%writefile max_pairwise_product_n2_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(i + 1, n):
            aux_product = l[i] * l[j]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_n2_py.py


ejecucion del cod fuente del algoritmo 2 en python

In [99]:
!python max_pairwise_product_n2_py.py < input11.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product_n2_py.py", line 17, in <module>
    print(max_pairwise_product(l))
  File "/content/max_pairwise_product_n2_py.py", line 9, in max_pairwise_product
    if max_product < aux_product:
KeyboardInterrupt
^C


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T (n) = 11 + 7n + \left[2(n-1)n+1, 2.5(n-1)n\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```

$$T(10) \in [262, 306] $$

In [100]:
n = 10
print(11 + 7*(10) + 2*(10)*(9) + 1, 11 + 7*(10) + 2.5*(10)*(9))

262 306.0


For the input11.txt
```
200000
1 2 . . . 199999 200000
```

$$T(200000) \in [80001000012, 100000900011] $$

In [101]:
print(11 + 7*(200000) + 2*(200000)*(199999) + 1, 11 + 7*(200000) + 2.5*(200000)*(199999))

80001000012 100000900011.0


In [209]:
#ins
estimated_time = 195 / 100000900009
estimated_time

1.9499824499824516e-09

In [103]:
#ins per sec (aprox.)
ins_per_sec =  1 / estimated_time
ins_per_sec

512825128.2512821

###Correctness tests

In [104]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long.cpp -o max_pairwise_product_long

for i in {0..10}
do
    ./max_pairwise_product_long < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [105]:
!./stress_test.sh

In [106]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_n2_py.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [107]:
!./stress_test.sh

6818678748
3361780485


###Estimated maximum that can be solved in 1 second
$$(11 + 7n + 2.5(n-1)n)*1.47*10^{-9} = 1s$$
$$ n \approx 16400 $$

In [108]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 16400 <<"\n";  // n
 for (int i=1; i <16401; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [109]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [110]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

268943600

real	0m0.919s
user	0m0.898s
sys	0m0.004s


In [111]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("16400\n")
        for i in range(1, 16401):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [112]:
!python WriteTestPython.py

In [113]:
#Slower. Not a surprise.
%%time
!python max_pairwise_product_n2_py.py < input14.txt

268943600
CPU times: user 163 ms, sys: 14.6 ms, total: 178 ms
Wall time: 19.9 s


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

algoritmo 3, el cual toma la lista de elementos y aplica un ordenamiento ascendente, por lo tanto, para obtener el producto mas grande debemos considerar los dos últimos valores de la lista.

In [114]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> v(n);                                   //n
    for(int i = 0; i < n; i++){                              //1 + 2n
        std::cin >> v[i];                                    //n
    }

    std::sort(v.begin(), v.end());                          //nlogn
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';   //2
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [115]:
%%writefile max_pairwise_product_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())                                        #1
    l = list(map(int, input().split()))                     #5n
    l.sort()                                                #nlogn at least
    print(l[-1] * l[-2])                              #2

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [116]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!./max_pairwise_product_sort < input1.txt   #Original array in input1.txt increasing order [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
!./max_pairwise_product_sort < input2.txt   #Original array in input2.txt decreasing order [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

90
90


In [117]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [118]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.003s
user	0m0.002s
sys	0m0.001s


In [119]:
%%time
#Given in secs. *10^-3 ms
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 6.78 ms, sys: 974 µs, total: 7.76 ms
Wall time: 111 ms


20 stress tests

In [120]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
      // obtener un valor aleatorio n para el tamaño de la lista entre 2 y 50
        int n = rand() % 49 + 2;
        std::vector<int> v(n);
      // obtenemos los n valores para llenar el arreglo (valores entre 2 y 1001)
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
      // ordenamiento del arreglo de manera ascendente
        std::sort(v.begin(), v.end());
      // multiplicacion del ultimo y penultimo valor
        std::cout << "test " << i+1 << ", list size, n: " << n << ", max product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [121]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

test 1, list size, n: 17, max product: 850976
test 2, list size, n: 25, max product: 857451
test 3, list size, n: 43, max product: 980036
test 4, list size, n: 10, max product: 692094
test 5, list size, n: 30, max product: 839816
test 6, list size, n: 48, max product: 956363
test 7, list size, n: 9, max product: 475580
test 8, list size, n: 46, max product: 954088
test 9, list size, n: 44, max product: 902850
test 10, list size, n: 17, max product: 981034
test 11, list size, n: 40, max product: 928812
test 12, list size, n: 7, max product: 267968
test 13, list size, n: 42, max product: 991020
test 14, list size, n: 27, max product: 908160
test 15, list size, n: 23, max product: 990016
test 16, list size, n: 2, max product: 141372
test 17, list size, n: 4, max product: 418500
test 18, list size, n: 11, max product: 557996
test 19, list size, n: 42, max product: 963270
test 20, list size, n: 17, max product: 690616


In [122]:
%%writefile stress_tests_sort_py.py
#Python3
import random
if __name__ == "__main__":
    for i in range(20):
        # generar el tamaño de la lista
        n = random.randint(2, 101)
        # llenar la lista con los n valores generados aleatoreos
        l = [ random.randint(10, 10001) for i in range(n) ]
        # ordenamos la lista de manera ascendente
        l.sort()
        print(f"test {i+1}, list size, n: {n}, max_product: {l[-1] * l[-2]}")

Writing stress_tests_sort_py.py


In [123]:
!python stress_tests_sort_py.py

test 1, list size, n: 30, max_product: 97306989
test 2, list size, n: 92, max_product: 98495100
test 3, list size, n: 73, max_product: 96478956
test 4, list size, n: 51, max_product: 93863760
test 5, list size, n: 4, max_product: 30861000
test 6, list size, n: 90, max_product: 98940873
test 7, list size, n: 27, max_product: 73213582
test 8, list size, n: 64, max_product: 98364788
test 9, list size, n: 18, max_product: 89052696
test 10, list size, n: 21, max_product: 80893890
test 11, list size, n: 56, max_product: 99040783
test 12, list size, n: 50, max_product: 90317745
test 13, list size, n: 84, max_product: 96606840
test 14, list size, n: 89, max_product: 97841730
test 15, list size, n: 83, max_product: 98444859
test 16, list size, n: 18, max_product: 93771392
test 17, list size, n: 97, max_product: 97651816
test 18, list size, n: 33, max_product: 92782690
test 19, list size, n: 51, max_product: 95558144
test 20, list size, n: 30, max_product: 92206584


Count executed instructions in code and verify with inputs 1,2 and 3


In [124]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                                                   //1
    ins++;
    std::cin >> n;                                           //1
    ins++;
    std::vector<uint64_t> v(n);                                   //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){                              //1 + 2n
        ins += 2;
        std::cin >> v[i];                                    //n
        ins++;
    }

    std::sort(v.begin(), v.end());                           //nlogn at least (nlogn time complexity)
    ins += n * log2(n);
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';    //3
    ins += 3;
    std::cout << "Executed instructions: " << ins << '\n';   // 7 + 4n + ins(nlogn sort). Where ins(nlogn sort) are, at least, nlogn.
    return 0;
}

Writing max_pairwise_product_sort_ins.cpp


In [125]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 79
90
Executed instructions: 79
40000000000
Executed instructions: 45


In [126]:
!time ./max_pairwise_product_sort_ins < input11.txt

39999800000
Executed instructions: 4321934

real	0m0.129s
user	0m0.122s
sys	0m0.003s


In [127]:
%%writefile max_pairwise_product_sort_ins.py
#Python3
import math

if __name__ == "__main__":
    ins = 0
    n = int(input())                       #1
    ins += 1
    l = list(map(int, input().split()))    #5n (n input read, n mappings to int, n iterable to list, n comparisons, n splits through input)
    ins += 5 * n
    l.sort()                               #nlogn at least
    ins += n * math.log2(n)
    print(l[-1] * l[-2])             #2
    ins += 2
    print(f"Executed instructions: {int(ins)}\n")                             #3 + 5n + ins(nlogn timsort) where ins(nlogn sort) are, at least, nlogn.

Writing max_pairwise_product_sort_ins.py


In [128]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Executed instructions: 86

90
Executed instructions: 86

40000000000
Executed instructions: 48



In [129]:
%%time
!python max_pairwise_product_sort_ins.py < input11.txt

39999800000
Executed instructions: 4521931

CPU times: user 11 ms, sys: 551 µs, total: 11.6 ms
Wall time: 207 ms


Calculate estimated time by instruction and number of instructions by second

In [130]:
#following result of input3.txt. #lower bound. Needs to know the exact number of instructions in each sorting algorithm. Just counts the ones in the nlogn step.
#Estimaded time by instruction and number of instructions
estimated_time = 0.009 / 4521931
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

1.9903001615902584e-09
502436777.77777785


###Estimated maximum that can be solved in 1 second
$$(5 + 4n + 5nlog(n))*1.47*10^{-9} = 1s \ \ \ given \ that \ nlogn \ at \ least$$
$$ n \approx 8'000.000 $$

In [131]:
# Verify input13.txt
%%writefile WriteTest.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 8000000 <<"\n";  // n
 for (int i=1; i <8000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [132]:
!g++ WriteTest.cpp -o WriteTest

In [133]:
!./WriteTest

In [134]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input13.txt

63999992000000

real	0m6.642s
user	0m6.476s
sys	0m0.035s


In [135]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input16.txt", "w") as f:
        f.write("8000000\n")
        for i in range(1, 8000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [136]:
!python WriteTestPython.py

In [137]:
#Finally..
%%time
!python max_pairwise_product_sort.py < input13.txt

63999992000000
CPU times: user 27.2 ms, sys: 4.2 ms, total: 31.4 ms
Wall time: 2.42 s


###Correctness Tests

In [138]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_sort.cpp -o max_pairwise_product_long

for i in {0..10}
do
    ./max_pairwise_product_sort < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [139]:
!./stress_test.sh

In [140]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_sort.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [141]:
!./stress_test.sh

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [142]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    long long p_max = 0;                     //1
    long long s_max = 0;                     //1
    for(int i = 0; i < n; ++i) {             //1 + 2n
        // actualizacion de los valores mas grande y segundo mas grande
        if (v[i] > s_max){                    //n
            s_max = v[i];                    //[0, n]
        }
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            p_max = v[i];                    //[0, n]
        }
    }
     // retornamos el producto de los dos valores encontrados
    return (long long) p_max * s_max;       //2 -> 9 + 4n + 3[0, n]           //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    // leemos por teclado el tamaño del arreglo
    std::cin >> n;                          //1
    std::vector<int> v(n);                  //n
    // guardamos en el arreglo los n valores solicitandolos por teclado
    for(int i = 0; i < n; i++){             //1 + 2n
        std::cin >> v[i];                   //n
    }
    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n
    return 0;
}

Writing max_pairwise_product_ef.cpp


In [143]:
%%writefile max_pairwise_product_ef.py
#Python3
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product_f(l))

Writing max_pairwise_product_ef.py


Test with inputs 1 and 2



In [144]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [145]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

90
90


Test and time it with input 3

In [146]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [147]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input11.txt

39999800000

real	0m0.134s
user	0m0.122s
sys	0m0.002s


In [148]:
%%time
!python max_pairwise_product_ef.py < input3.txt

40000000000
CPU times: user 6.83 ms, sys: 0 ns, total: 6.83 ms
Wall time: 110 ms


In [149]:
%%time
!python max_pairwise_product_ef.py < input11.txt

39999800000
CPU times: user 9.17 ms, sys: 0 ns, total: 9.17 ms
Wall time: 310 ms


20 stress tests

In [150]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();
    ins++;
    long long p_max = 0;
    ins++;
    long long s_max = 0;
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {
        ins += 2;
        ins++;
        if (v[i] > s_max){
            s_max = v[i];
            ins++;
        }
        ins++;
        if (v[i] > p_max){
            s_max = p_max;
            ins++;
            p_max = v[i];
            ins++;
        }
    }

    return (long long) p_max * s_max;
}

int main(int argc, char* argv[]){

  // 20 casos de prueba
    for(int i = 0; i < 20; i++){
      // tamaño de la lista entre 2 - 70
        int n = rand() % 69 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
          // llenado del arreglo con n valores aleatorios entre 2 - 1001
            v[i] = rand() % 1000 + 2;
        }
        std::cout << "n: " << n << " Product: " << max_pairwise_product_f(v) << '\n';
    }
    return 0;
}

Writing stress_tests_ef_cpp.cpp


In [151]:
%%writefile stress_tests_ef_py.py
#Python3
import random
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        l = [random.randint(17, 24531) for i in range(n)]
        print(f"n: {n} Product: {max_pairwise_product_f(l)}")

Writing stress_tests_ef_py.py


In [152]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 36 Product: 863968
n: 57 Product: 980036
n: 60 Product: 898343
n: 31 Product: 927682
n: 13 Product: 762600
n: 51 Product: 954088
n: 12 Product: 683388
n: 26 Product: 981034
n: 70 Product: 960364
n: 14 Product: 858552
n: 61 Product: 993010
n: 52 Product: 963270
n: 10 Product: 631764
n: 32 Product: 994008
n: 21 Product: 960319
n: 24 Product: 828625
n: 2 Product: 74562
n: 54 Product: 904397
n: 8 Product: 886974
n: 23 Product: 911064


In [153]:
!python stress_tests_ef_py.py

n: 91 Product: 553241676
n: 56 Product: 582234099
n: 10 Product: 425054877
n: 13 Product: 533036700
n: 95 Product: 581078211
n: 33 Product: 581820525
n: 68 Product: 582976536
n: 69 Product: 593165574
n: 71 Product: 575950240
n: 77 Product: 553743360
n: 11 Product: 500543043
n: 44 Product: 593569002
n: 92 Product: 594756319
n: 10 Product: 593230720
n: 2 Product: 189065274
n: 49 Product: 581566486
n: 39 Product: 540274028
n: 12 Product: 386660153
n: 7 Product: 529533035
n: 34 Product: 470542464


Count executed instructions in code and verify with inputs 1,2 and 3


In [154]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    ins++;
    long long p_max = 0;                     //1
    ins++;
    long long s_max = 0;                     //1
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {             //1 + 2n
        ins += 2;
        ins++;
        if (v[i] > s_max){                    //n
            s_max = v[i];                     //[0, n]
            ins++;
        }
        ins++;
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            ins++;
            p_max = v[i];                    //[0, n]
            ins++;
        }
    }
    ins += 2;
    return (long long) p_max * s_max;       //2 -> 6 + 4n + 3[0, n]          //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    ins++;
    std::cin >> n;                          //1
    ins++;
    std::vector<int> v(n);                  //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){             //1 + 2n
        ins += 2;
        std::cin >> v[i];                   //n
        ins++;
    }

    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n + 6 + 4n + 3[0, n] -> 10 +8n + 3[0, n] -> 10 + 11n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_ef_ins.cpp


In [155]:
%%writefile max_pairwise_product_ef_ins_py.py
#Python3
ins = 0
def max_pairwise_product_f(l):
    global ins
    p_max = 0                               #1
    ins += 1
    s_max = 0                               #1
    ins += 1
    ins += 1
    for i in l:                             #1 + 2n
        ins += 2
        if i > s_max:                       #n
            s_max = i                       #[0, n]           #[0, n] means that, in the worst case, executes n times
            ins += 1
        if i > p_max:                       #n
            s_max = p_max                   #[0, n]
            ins += 1
            p_max = i                       #[0, n]
            ins += 1

    ins += 2
    return p_max * s_max                    #2       #5 + 4n + 3[0, n] -> 5 + 7n worst

if __name__ == "__main__":
        n = int(input())                                 #1
        ins += 1
        l = list(map(int, input().split()))              #5n
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))                 #1
        print(f"Executed instructions: {ins}")           #-> 2 + 5n + 5 + 7n -> 7 + 12n

Writing max_pairwise_product_ef_ins_py.py


In [156]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 119
90
Executed instructions: 93
40000000000
Executed instructions: 68


In [157]:
!time ./max_pairwise_product_ef_ins < input11.txt

39999800000
Executed instructions: 2200009

real	0m0.081s
user	0m0.077s
sys	0m0.002s


In [158]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


In [159]:
%%time
!python max_pairwise_product_ef_ins_py.py < input3.txt

40000000000
Executed instructions: 60
CPU times: user 8.4 ms, sys: 0 ns, total: 8.4 ms
Wall time: 112 ms


In [160]:
%%time
!python max_pairwise_product_ef_ins_py.py < input11.txt

39999800000
Executed instructions: 2000007
CPU times: user 9.03 ms, sys: 0 ns, total: 9.03 ms
Wall time: 205 ms


Calculate estimated time by instruction and number of instructions by second

In [161]:
estimated_time = 0.006/ 2200009
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

2.727261570293576e-09
366668166.6666667


###Correctness Tests

script para probar la salida esperada con la salida obtenida en el algoritmo eficiente en c++

In [162]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

for i in {0..10}
do
    ./max_pairwise_product_ef_cpp < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [163]:
!./stress_test.sh

script para probar la salida esperada con la salida obtenida en el algoritmo eficiente en python

In [164]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_ef.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [165]:
!./stress_test.sh

6818678748
3361780485


###Estimated maximum that can be solved in 1 second
$$(10 + 11n)*1.47*10^{-9} = 1s$$
$$ n \approx 60'000.000 $$

In [166]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 60000000 <<"\n";  // n
 for (int i=1; i <60000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [167]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [168]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

In [169]:
!time ./max_pairwise_product_ef_cpp < input15.txt #This gives the "practical" time instead of the theoretical one (1s). The "real" constants are way bigger.

3599999940000000

real	0m26.698s
user	0m26.014s
sys	0m0.346s


In [170]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("60000000\n")
        for i in range(1, 60000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [171]:
!python WriteTestPython.py

In [172]:
%%time
!python max_pairwise_product_ef.py < input14.txt #This gives the "practical" time instead of the theoretical one (1s). The "real" constants are way bigger.

3599999940000000
CPU times: user 278 ms, sys: 34.6 ms, total: 312 ms
Wall time: 29.9 s


### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [173]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    int n;
    std::cin >> n;
    int p_max = INT_MIN;
    int s_max = INT_MIN;

    for (int i = 0; i < n; i++){
        int x;
        std::cin >> x;
        // leemos el dato y acto seguido validamos su tamaño
        if(x > p_max){
          // guardamos el valor de p_max como el segundo mas grande
            s_max = p_max;
          // guardamos el valor de x como el primero mas grande
            p_max = x;
        }
        // x no es mas grande que p_max pero si mas grande que s_max
        else if(x > s_max){
          // actualizacion del segundo mas grande
            s_max = x;
        }
    }

    // mostramos en pantalla en producto de los dos valores mas grandes encontrados

    std::cout << (long long) p_max * s_max << "\n";
    return 0;
}

Writing max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [174]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [175]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [176]:
!time ./max_pairwise_product_gef_cpp < input11.txt

39999800000

real	0m0.072s
user	0m0.058s
sys	0m0.006s


20 stress tests

In [177]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){

  // 20 casos de prueba
    for(int i = 0; i < 20; i++){

      // tamaño de la lista definido desde 2 hasta 78
        int n = rand() % 77 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

      // generacion de los n valores y comparacion con p_max y s_max
        for (int i = 0; i < n; i++){
          // valores generados desde 2 hasta 1001
            int x = rand() % 1000 + 2;
          // comparacion con el primero mas grande
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
          // comparacion con el segundo mas grande
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Writing stress_tests_gef_cpp.cpp


In [178]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 52 Product: 863968
n: 59 Product: 956084
n: 53 Product: 898343
n: 22 Product: 927682
n: 2 Product: 56643
n: 32 Product: 928140
n: 48 Product: 885256
n: 26 Product: 981034
n: 40 Product: 928812
n: 23 Product: 876600
n: 55 Product: 991020
n: 27 Product: 990016
n: 11 Product: 557996
n: 70 Product: 994008
n: 53 Product: 961308
n: 58 Product: 926316
n: 39 Product: 979080
n: 3 Product: 290360
n: 23 Product: 876015
n: 26 Product: 874290


Count executed instructions in code and verify with inputs 1,2 and 3

In [179]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                          //1
    ins++;
    std::cin >> n;                  //1
    ins++;
    int p_max = INT_MIN;            //1
    ins++;
    int s_max = INT_MIN;            //1
    ins++;
    ins++;
    for (int i = 0; i < n; i++){    //1 + 2n
        ins += 2;
        int x;                      //n
        ins++;
        std::cin >> x;              //n
        ins++;
        ins += 2; //both ifs
        if(x > p_max){              //[0, n]           //[0, n] means that, in the worst case, executes n times.
            s_max = p_max;          //[0, n]
            ins++;
            p_max = x;              //[0, n]
            ins++;
        }
        else if(x > s_max){         //[0, n]
            s_max = x;              //[0, n]
            ins++;
        }
    }

    ins += 3;
    std::cout << (long long) p_max * s_max << "\n"; //3 -> 8 + 4n + 5[0, n] -> 8 + 9n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_gef_ins_cpp.cpp


In [180]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!./max_pairwise_product_gef_ins_cpp < input1.txt
!./max_pairwise_product_gef_ins_cpp < input2.txt
!./max_pairwise_product_gef_ins_cpp < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52


In [181]:
!time ./max_pairwise_product_gef_ins_cpp < input11.txt

39999800000
Executed instructions: 1600008

real	0m0.072s
user	0m0.062s
sys	0m0.006s


Calculate estimated time by instruction and number of instructions by second

In [182]:
estimated_time = 0.0071 / 1600008
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.437477812610937e-09
225353239.4366197


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [183]:
%%writefile max_pairwise_product_gef_py.py
#Python3
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        x = int(i)
        if(x > p_max):
            s_max = p_max
            p_max = x
        elif x > s_max:
            s_max = x
    print(p_max * s_max)

Writing max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [184]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [185]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 7.18 ms, sys: 0 ns, total: 7.18 ms
Wall time: 108 ms


In [186]:
%%time
!python max_pairwise_product_gef_py.py < input11.txt

39999800000
CPU times: user 8.29 ms, sys: 111 µs, total: 8.4 ms
Wall time: 204 ms


20 stress tests

In [187]:
%%writefile stress_tests_gef_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        p_max = -1
        s_max = -1
        for x in range(n):
            i = random.randint(19, 133700)
            if(i > p_max):
                s_max = p_max
                p_max = i
            elif x > s_max:
                s_max = x
        max_product_super_ef =p_max * s_max
        print(f"n: {n} Product: {p_max * s_max}")

Writing stress_tests_gef_py.py


In [189]:
!python stress_tests_gef_py.py

n: 86 Product: 16825721904
n: 74 Product: 16722887986
n: 92 Product: 16833224286
n: 21 Product: 17420787051
n: 57 Product: 17402412083
n: 41 Product: 17256514289
n: 26 Product: 16771343144
n: 2 Product: 75073
n: 92 Product: 15636443724
n: 58 Product: 17223162978
n: 77 Product: 17116904352
n: 89 Product: 15546928584
n: 40 Product: 17430961366
n: 9 Product: 14151129096
n: 5 Product: 16733950064
n: 59 Product: 7321450812
n: 7 Product: 7633066728
n: 54 Product: 3675611674
n: 14 Product: 1707316
n: 11 Product: 16452071904


Count executed instructions in code and verify with inputs 1,2 and 3

In [190]:
%%writefile max_pairwise_product_gef_ins_py.py
import random
#Python3
if __name__ == "__main__":
        ins = 0
        n = int(input())                 #1
        ins += 1
        p_max = -1                       #1
        ins += 1
        s_max = -1                       #1
        ins += 1
        ins += 1
        for i in input().split():        #1 + 3n (n reads, 2n comparisons unitl end of input)
            ins += 2
            i = int(i)                   #n
            ins += 1
            ins += 2
            if(i > p_max):               #[0, n]
                s_max = p_max            #[0, n]
                ins += 1
                p_max = i                #[0, n]
                ins += 1
            elif i > s_max:              #[0, n]
                s_max = i                #[0, n]
                ins += 1
        print(p_max * s_max)             #2
        print(f"Executed instructions: {ins}") #->6 + 4n + 5[0, n] -> 6 + 9n worst case

Writing max_pairwise_product_gef_ins_py.py


In [191]:
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 74
90
Executed instructions: 57
40000000000
Executed instructions: 42


In [192]:
%%time
!python max_pairwise_product_gef_ins_py.py < input11.txt

39999800000
Executed instructions: 1400004
CPU times: user 7.32 ms, sys: 144 µs, total: 7.46 ms
Wall time: 305 ms


Calculate estimated time by instruction and number of instructions by second

In [193]:
estimated_time = 0.0066 / 1400004
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.714272244936443e-09
212121818.1818182


###Correctness tests

In [194]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp

for i in {0..10}
do
    ./max_pairwise_product_gef_cpp < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [195]:
!./stress_test.sh

In [196]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_gef_py.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [197]:
!./stress_test.sh

6818678748
3361780485


###Estimated maximum that can be solved in 1 second
$$(6 + 9n)*1.47*10^{-9} = 1s$$
$$ n \approx 75'000.000 $$

In [198]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 75000000 <<"\n";  // n
 for (int i=1; i <75000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [199]:
!g++ WriteTest.cpp -o WriteTest

In [200]:
!./WriteTest

In [201]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input15.txt #The theoretical max in 1 sec is 75M. The constants are larger in practice.

5624999925000000

real	0m32.099s
user	0m31.607s
sys	0m0.257s


In [202]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("75000000\n")
        for i in range(1, 75000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [203]:
!python WriteTestPython.py

In [204]:
%%time
!python max_pairwise_product_gef_py.py < input14.txt #Again, the theoretical holds. The constans are larger in practice.

5624999925000000
CPU times: user 298 ms, sys: 54.8 ms, total: 353 ms
Wall time: 40.6 s


## Perform 100 randomized automated tests in C++ and Python between the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n ln g))$.

100 stress tests python


In [205]:
%%writefile _100_automated_tests.py
#Python3
import random
if __name__ == "__main__":
    for i in range(100):
        n = random.randint(2, 10001)
        p_max = 0
        s_max = 0
        arr = []

        for j in range(n):
          x = random.randint(2, 10001)
          arr.append(x)

          if(x > p_max):
            s_max = p_max
            p_max = x
          elif x > s_max:
            s_max = x

        arr.sort()
        max_product_sort = arr[-1] * arr[-2]
        max_product_super_efficient = p_max * s_max

        if max_product_sort != max_product_super_efficient:
          print("ERROR")
          print(f"Test: {i + 1}, n: {n} ")
          print(f"max_product_sort: {max_product_sort}")
          print(f"max_product_super_efficient: {max_product_super_efficient}")
          print(f"lista: {arr}")

          exit()

    print("PASS")

Writing _100_automated_tests.py


In [206]:
!python _100_automated_tests.py

PASS


100 stress tests algoritmos en c++




In [207]:
%%writefile _100_automated_tests_.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){


    // 100 casos de prueba
    for(int i = 0; i < 100; i++){

        //declaracion de var. para guardar el primer y segundo mas grande
        int p_max = 0;
        int s_max = 0;

        /* para el algoritmo O(nlgn)
         generamos un valor n aleatorio para el tamaño de la lista
         n varia entre 2 - 10001 */
        int n = rand() % 10000 + 2;

        /* para el algo. O(nlgn)  declaramos el arreglo de tamaño n */
        std::vector<int64_t> v(n);


        for(int j = 0; j < n; j++){
            /* para el algo. O(nlgn) llenamos el arreglo con n valores aleatorios
             * entre 2 - 10001 */
            int x = rand() % 10000 + 2;
            v[j] = x;

            /*para el algo. O(n) comparamos el valor x generado con p_max s_max*/
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }
        //ordenamos el arreglo
        std::sort(v.begin(), v.end());
        // calculamos los productos obtenidos
        long long max_product_sort = v[n - 1] * v[n - 2];
        long long max_product_super_ef = p_max * s_max;

        // comparacion de resultados
        if (max_product_sort != max_product_super_ef){
            // error
            std::cout << "ERROR "<< '\n';
            std::cout << "Test: " << i + 1 << " n: " << n << '\n';
            std::cout << "max_product_super_ef: " << (long long) max_product_super_ef << '\n';
            std::cout << "max_product_sort: " << (long long) max_product_sort << '\n';

            return 0;

        }
    }
    std::cout << "PASS "<< '\n';
    return 0;
}

Writing _100_automated_tests_.cpp


In [208]:
!g++ _100_automated_tests_.cpp -o _100_automated_tests_
!./_100_automated_tests_

PASS 
